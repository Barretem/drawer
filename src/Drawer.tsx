import * as React from 'react';
import type { GetContainer } from 'rc-util/lib/PortalWrapper';
import Portal from 'rc-util/lib/PortalWrapper';
import DrawerPopup from './DrawerPopup';
import type { DrawerPopupProps } from './DrawerPopup';

export type Placement = 'left' | 'top' | 'right' | 'bottom';

type LevelMove = number | [number, number];

export interface DrawerProps extends Omit<DrawerPopupProps, 'prefixCls'> {
  prefixCls?: string;

  width?: string | number;
  height?: string | number;
  open?: boolean;
  handler?: React.ReactElement | null | false;
  placement?: Placement;
  level?: null | string | string[];
  levelMove?:
    | LevelMove
    | ((e: { target: HTMLElement; open: boolean }) => LevelMove);
  duration?: string;
  ease?: string;
  showMask?: boolean;
  onChange?: (open?: boolean) => void;
  onHandleClick?: (e: React.MouseEvent | React.KeyboardEvent) => void;
  onClose?: (e: React.MouseEvent | React.KeyboardEvent) => void;
  keyboard?: boolean;
  contentWrapperStyle?: React.CSSProperties;
  autoFocus?: boolean;
  wrapperClassName?: string;
  forceRender?: boolean;
  getContainer?: GetContainer | false;
}

const defaultGetContainer = () => document.body;

export default function Drawer(props: DrawerProps) {
  const {
    open,
    getContainer = defaultGetContainer,
    forceRender,
    wrapperClassName,
    prefixCls = 'rc-drawer',
  } = props;

  const sharedDrawerProps = {
    ...props,
    prefixCls,
  };

  if (getContainer === false) {
    return <DrawerPopup {...sharedDrawerProps} inline />;
  }

  return (
    <Portal
      visible={open}
      forceRender={forceRender}
      getContainer={getContainer}
      wrapperClassName={wrapperClassName}
    >
      {({ scrollLocker }) => (
        <DrawerPopup {...sharedDrawerProps} scrollLocker={scrollLocker} />
      )}
    </Portal>
  );
}
