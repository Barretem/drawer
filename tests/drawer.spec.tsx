import React, { useRef } from 'react';
import Drawer from '../src/';
import type { IDrawerProps } from '../src/IDrawerPropTypes';
import { cleanup, render } from '@testing-library/react';

const DrawerTesterRef = (props: IDrawerProps) => {
  const containerRef = useRef<HTMLDivElement>(null);
  return (
    <div>
      <div ref={containerRef} className="main" />
      <Drawer getContainer={() => containerRef.current} {...props}>
        <p className="text">Here is content of Drawer</p>
      </Drawer>
    </div>
  );
};

const DrawerTesterDom = (props: IDrawerProps) => {
  const containerRef = useRef<HTMLDivElement>(null);
  return (
    <div>
      <div ref={containerRef} className="main" />
      {open && (
        <Drawer getContainer={() => containerRef.current} {...props}>
          <p className="text">Here is content of Drawer</p>
        </Drawer>
      )}
    </div>
  );
};

describe('Drawer', () => {
  afterEach(() => {
    cleanup();
  });
  it('render function', () => {
    render(<DrawerTesterRef open />);
    expect(document.body.children[0]).toMatchSnapshot();
  });

  it('render dom', () => {
    const { rerender } = render(<DrawerTesterDom />);
    rerender(<DrawerTesterDom open />);
    expect(document.body.children[0]).toMatchSnapshot();
  });

  it('render boolean', () => {
    render(<DrawerTesterRef open getContainer={false} />);
    expect(document.body.children[0]).toMatchSnapshot();
  });
});
