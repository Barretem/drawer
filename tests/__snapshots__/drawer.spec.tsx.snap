// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drawer render boolean 1`] = `
<body>
  <div>
    <div>
      <div
        class="main"
        style="transition: transform .3s cubic-bezier(0.78, 0.14, 0.15, 0.86);"
      />
      <div
        class=""
      >
        <div
          class="drawer drawer-left drawer-open"
          tabindex="-1"
        >
          <div
            class="drawer-mask"
          />
          <div
            class="drawer-content-wrapper"
            style=""
          >
            <div
              class="drawer-content"
            >
              <p
                class="text"
              >
                Here is content of Drawer
              </p>
            </div>
            <div
              class="drawer-handle"
            >
              <i
                class="drawer-handle-icon"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Drawer render dom 1`] = `
<body>
  <div>
    <div>
      <div
        class="main ant-scrolling-effect"
        style="overflow: hidden; overflow-x: hidden; overflow-y: hidden;"
      >
        <div>
          <div
            class="drawer drawer-left drawer-open"
            tabindex="-1"
          >
            <div
              class="drawer-mask"
            />
            <div
              class="drawer-content-wrapper"
              style=""
            >
              <div
                class="drawer-content"
              >
                <p
                  class="text"
                >
                  Here is content of Drawer
                </p>
              </div>
              <div
                class="drawer-handle"
              >
                <i
                  class="drawer-handle-icon"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Drawer render function 1`] = `
<body>
  <div>
    <div>
      <div
        class="main ant-scrolling-effect"
        style="overflow: hidden; overflow-x: hidden; overflow-y: hidden;"
      >
        <div>
          <div
            class="drawer drawer-left drawer-open"
            tabindex="-1"
          >
            <div
              class="drawer-mask"
            />
            <div
              class="drawer-content-wrapper"
              style=""
            >
              <div
                class="drawer-content"
              >
                <p
                  class="text"
                >
                  Here is content of Drawer
                </p>
              </div>
              <div
                class="drawer-handle"
              >
                <i
                  class="drawer-handle-icon"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
